### PR TITLE
Merge the Sample size calculation feature for dose response curve into Development

### LINKED ISSUE #167: [Sample-Size] Integrate sample size calculation into MSstatsShiny
## Story

Experimentalists need to understand how many samples they need to reproduce results for a future experiment.  Typically, people will run pilot studies with a few samples, and then assess variation, and use that variation to run a real larger study.  But more samples = money

## Subtasks
- Scope the [sample size UI](https://github.com/Vitek-Lab/MSstatsShiny/blob/c0b2ee5cfd33ba5484bf6a73e151c1768dbfa240/R/module-expdes-ui.R) and [sample size server](https://github.com/Vitek-Lab/MSstatsShiny/blob/c0b2ee5cfd33ba5484bf6a73e151c1768dbfa240/R/module-expdes-server.R) to understand how to best not repeat bad patterns.
- Create two functions.  One function generates the data table of simulation results.  The second function generates a plot based on the simulation results.  See example script below.
- If dose response curve option is enabled, then call the new functions when encountering the sample size page.  

## Script

```
# Create interaction detection power curves using simulations --

library(MSstatsResponse)
library(tidyverse)
library(data.table)


temp_res = futureExperimentSimulation(N_proteins = 300,
                           N_rep = 4,
                           Concentrations = c(0, 1, 1000, 3000),
                           IC50_Prediction = FALSE)
temp_res$Hit_Rates_Data


# 2 con : 0, 3000
# 3 con : 0, 1000, 3000,
# 4 con : 0, 1, 1000, 3000
# 5 con : 0, 1, 100, 1000, 3000
# 6 concentration : 0, 1, 100, 300, 1000, 3000
# 7 concentration : 0, 1, 10, 100, 300, 1000, 3000
# 8 concentration : 0, 1, 10, 30, 100, 300, 1000, 3000
# 9 concentration : 0, 1, 3, 10, 30, 100, 300, 1000, 3000


# ---- Setup -------------------------------------------------------------------
library(dplyr)
library(purrr)
library(tidyr)
library(ggplot2)

# Allowed concentrations (these are the ones that exist in the experiment)
allowed_concs <- c(0, 1, 3, 10, 30, 100, 300, 1000, 3000)

# custom  selection for each number of concentrations
conc_map <- list(
  `2` = c(0, 3000),
  `3` = c(0, 1000, 3000),
  `4` = c(0, 1, 1000, 3000),
  `5` = c(0, 1, 100, 1000, 3000),
  `6` = c(0, 1, 100, 300, 1000, 3000),
  `7` = c(0, 1, 10, 100, 300, 1000, 3000),
  `8` = c(0, 1, 10, 30, 100, 300, 1000, 3000),
  `9` = c(0, 1, 3, 10, 30, 100, 300, 1000, 3000)
)

# Grid to sweep: replicates 1–4 and K in {2..9}
k_grid   <- as.integer(names(conc_map))
#rep_grid <- 1:4
rep_grid <- 1:5

# Helper to run one config mapping
run_one <- function(n_rep, k_conc, seed = 123) {
  set.seed(seed + n_rep * 100 + k_conc)

  concs_k <- conc_map[[as.character(k_conc)]]
  if (is.null(concs_k)) stop("No concentration set defined for K = ", k_conc)
  if (!all(concs_k %in% allowed_concs)) {
    stop("Concentration set for K=", k_conc, " includes values outside allowed_concs.")
  }

  temp_res <- futureExperimentSimulation(
    N_proteins      = 300,
    N_rep           = n_rep,
    Concentrations  = concs_k,
    IC50_Prediction = FALSE
  )

  temp_res$Hit_Rates_Data %>%
    dplyr::filter(Category %in% c("TPR (Strong)", "TPR (Weak)")) %>%
    dplyr::mutate(
      N_rep      = n_rep,
      NumConcs   = length(concs_k),
      Interaction = if_else(Category == "TPR (Strong)", "Strong", "Weak")
    ) %>%
    dplyr::select(Interaction, TPR = Percent, N_rep, NumConcs)
}

# Run  full 
grid_df <- expand.grid(N_rep = rep_grid, k_conc = k_grid) %>% as_tibble()
results_long <- pmap_dfr(grid_df, ~ run_one(n_rep = ..1, k_conc = ..2))

# visualize 
results_strong = results_long %>% filter(Interaction == 'Strong')
results_weak = results_long %>% filter(Interaction == 'Weak')

p_power <- ggplot(results_weak,
                  aes(x = NumConcs, y = TPR,
                      color = Interaction, linetype = factor(N_rep))) +
  geom_line(linewidth = 1.5) +
  geom_point(size = 2) +
  scale_x_continuous(breaks = k_grid) +
  scale_y_continuous(limits = c(0, 100), breaks = seq(0, 100, 20)) +
  scale_color_manual(values = c("Strong" = "#1b9e77", "Weak" = "#d95f02")) +
  #scale_linetype_manual(values = c("1" = "solid", "2" = "dashed", "3" = "dotdash", "4" = "dotted")) +
  scale_linetype_manual(values = c("1" = "dotted", "2" = "dotdash", "3" = "dashed", "4" = "longdash", "5" = "solid")) +

  labs(
    title = "Interaction Detection Power Curve",
    #subtitle = "Lines = Strong vs. Weak; Line type = Replicates per dose (1–4)",
    x = "Number of concentrations ",
    y = "True Positive Rate (%)",
    color = "Interaction",
    linetype = "Replicates"
  ) +
  theme_bw(base_size = 16) +
  theme(
    plot.title = element_text(face = "bold", hjust = 0.5),
    legend.position = "right"
  )

p_power

```

### PR DESCRIPTION
**Pull request by swaraj-neu**
- Integrate the sample size calculation into MSstatsShiny
- Adjust the max replicates per dose to 5 in future experiments
- Update the TPR simulation docs
- Add try-catch block in TRP simulation function for better error handling

## Motivation & Context

This PR integrates a dose-response (response-curve) sample size / power calculation workflow into MSstatsShiny by reusing simulation utilities from MSstatsResponse. It adds a response-curve mode that lets users select a protein, choose replicates per dose (capped at 5), run TPR simulations, and view/download power curves. The change centralizes TPR simulation logic (imported from MSstatsResponse), improves error handling around simulations (try-catch), and adapts UI/server wiring so both standard sample-size and dose-response workflows coexist.

## Detailed Changes

- Namespace & imports
  - Imported MSstatsResponse functions: futureExperimentSimulation, run_tpr_simulation, plot_tpr_power_curve (NAMESPACE and Roxygen updates).

- New constants
  - Added exported constant NAMESPACE_EXPDES: list(sidebar_controls, protein_select, rep_range, run_simulation, result_plot, download_future) for module UI namespacing.

- expdes module — server
  - expdesServer signature extended: added optional arguments preprocess_data = NULL and statmodel_contrast = NULL.
  - Added reactive prepared_response_data(): merges preprocess_data()$ProteinLevelData with statmodel_contrast$matrix on GROUP and prepares dose-response fit inputs.
  - Added response-curve mode detection (based on statmodel inputs) and mode-dependent behavior.
  - Added simulation workflow:
    - ui-driven replicates-per-dose slider (max capped at 5) and "Run simulation" button in response-curve mode.
    - observeEvent for Run simulation: extracts concentration/dose values from statmodel_contrast$matrix, validates presence, invokes run_tpr_simulation(...) with selected protein and replicate range, stores results in a reactive simulation_results.
    - UI feedback around simulation (modal spinner) and try-catch around run_tpr_simulation to notify users on failures.
  - Plots & downloads in response-curve mode:
    - result_plot uses plot_tpr_power_curve(simulation_results()).
    - Download handler generates a PDF showing interaction detection power curves or a friendly "run simulation first" message if results are missing.
  - Removed the prior unconditional future_exp() → plot/download pipeline and gated standard-mode rendering behind !is_response_curve().

- expdes module — UI
  - Replaced hard-coded sidebar inputs with a uiOutput that renders mode-appropriate controls.
  - Simplified explanatory text; wired main-panel outputs using NAMESPACE_EXPDES IDs.

- statmodel module & server integration
  - statmodelServer: returned list now exposes contrast = contrast reactive.
  - server.R: extracts statmodel_contrast from statmodelServer output and passes preprocess_data and statmodel_contrast into expdesServer via callModule.

- Documentation
  - man/expdesServer.Rd: updated to document the new optional parameters preprocess_data and statmodel_contrast and describe the contrast matrix usage.
  - Updated TPR simulation documentation and plot color palette (commit note).

- Miscellaneous
  - constants.R: fixed list closing and exported NAMESPACE_EXPDES.
  - Small UI text cleanup and whitespace/trailing-newline notes.
  - Clear previous sample size calculation results at start of each run (UX nitpick fix).

## Unit Tests

- No unit tests were added or modified in this PR.
  - There is no automated coverage for:
    - expdesServer new parameters and response-curve workflows,
    - prepared_response_data merging behavior and edge cases,
    - run_tpr_simulation integration and the new try-catch error handling,
    - the response-curve PDF download handler and plot generation.

## Coding Guidelines Violations

- Hardcoded values remain:
  - Maximum replicates per dose fixed at 5 (UI slider limit and logic).
  - PDF output dimensions and plot color codes appear hardcoded.
- Inline UI strings not externalized (labels/messages embedded in code).
- Insufficient input validation / defensive checks:
  - prepared_response_data() assumes successful merge; no explicit checks for empty or invalid merge results before simulation.
  - Minimal validation beyond UI constraints for rep_range and protein selection prior to invoking run_tpr_simulation.
- Minor repository hygiene:
  - Some files lack a trailing newline.
- Note: Error handling was improved by adding a try-catch around simulation invocation, but further structured error propagation and unit tests for failure modes are still recommended.

### CODE REVIEW

File: man/run_tpr_simulation.Rd [hunk 1]
```diff
@@ -0,0 +1,19 @@
+% Generated by roxygen2: do not edit by hand
+% Please edit documentation in R/module-expdes-server.R
+\name{run_tpr_simulation}
+\alias{run_tpr_simulation}
+\title{Run TPR simulation across a grid of concentration counts and replicate counts}
+\usage{
+run_tpr_simulation(rep_range, n_proteins = 1000)
+}
+\arguments{
+\item{rep_range}{integer vector of length 2, c(min, max) for replicate sweep}
```
Comment (tonywu1999) [inline man/run_tpr_simulation.Rd]: Users should also pass in the actual dataset and the Protein ID that is considered a strong interaction.

File: R/module-expdes-server.R [hunk 2]
```diff
@@ -1,7 +1,142 @@
...
+  "2" = c(0, 3000),
+  "3" = c(0, 1000, 3000),
+  "4" = c(0, 1, 1000, 3000),
+  "5" = c(0, 1, 100, 1000, 3000),
+  "6" = c(0, 1, 100, 300, 1000, 3000),
+  "7" = c(0, 1, 10, 100, 300, 1000, 3000),
+  "8" = c(0, 1, 10, 30, 100, 300, 1000, 3000),
+  "9" = c(0, 1, 3, 10, 30, 100, 300, 1000, 3000)
+)
+
```
Comment (tonywu1999) [inline R/module-expdes-server.R]: As discussed with Sarah, this should not be hard coded but rather each subsequent value should be picked based on farthest distance from the log(median).

File: R/module-expdes-server.R [hunk 3]
```diff
@@ -11,64 +12,232 @@
...
+            ggplot2::scale_color_manual(values = color_palette) +
+            ggplot2::labs(title = "Interaction detection power",
+                         x = "Number of concentrations",
+                         y = "True Positive Rate (%)", color = "Replicates") +
+            ggplot2::theme_bw(base_size = 14) +
+            ggplot2::theme(plot.title = ggplot2::element_text(face = "bold", hjust = 0.5))
+
+          pdf(file, width = 10, height = 6)
+          print(p)
+          dev.off()
```
Comment (tonywu1999) [inline R/module-expdes-server.R]: I think this is already behind an MSstatsResponse function?  The download handler should call the MSstatsresponse plotting function.

File: R/module-expdes-server.R [hunk 4]
```diff
@@ -11,64 +12,232 @@
...
+    }, error = function(e) {
+      remove_modal_spinner()
+      showNotification(conditionMessage(e), type = "error", duration = 8)
     })
   })
 
-}
+  # ---- Standard sample size logic ----
+  observe({
+    if (!is_response_curve()) {
```
Comment (tonywu1999) [inline R/module-expdes-server.R:185]: Could you refactor the code to be more friendly to unit tests?  Right now, there's no unit tests that validate that the new and existing behavior is working.  

For unit testing, just sanity check that certain functions are called whether it's with dose response curves or the default.

File: tests/testthat/test-module-expdes-server.R [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+test_that("is_response_curve returns FALSE for NULL input", {
+  mock_statmodel_input <- NULL
+
+  is_rc <- !is.null(mock_statmodel_input) &&
+    !is.null(mock_statmodel_input[[NAMESPACE_STATMODEL$comparison_mode]]) &&
+    mock_statmodel_input[[NAMESPACE_STATMODEL$comparison_mode]] ==
+      CONSTANTS_STATMODEL$comparison_mode_response_curve
+
+  expect_false(is_rc)
+})
```
Comment (tonywu1999) [inline tests/testthat/test-module-expdes-server.R]: These unit tests don't test is_response_curve directly.  If someone changes the logic of the is_response_curve reactive, these tests will still pass, defeating the purpose of these unit tests.  

I recommend refactoring the code into helper functions and test those helper functions.

**Review by tonywu1999 — APPROVED** [review]
